File: basic-proxy-impl/src/test/java/com/test/util/PrettyResultBuilder.java
package com.test.util;

import org.objectweb.asm.ClassReader;
import org.objectweb.asm.util.TraceClassVisitor;

import java.io.ByteArrayOutputStream;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class PrettyResultBuilder {

    private final boolean debug;

    public PrettyResultBuilder(boolean debug) {
        this.debug = debug;
    }

    public List<List<String>> build(byte[] raw) {
        OutputStream outputStream = new ByteArrayOutputStream();

        new ClassReader(raw).accept(new TraceClassVisitor(new PrintWriter(outputStream)), 0);
        List<String> lines =
                Arrays.stream(outputStream.toString().split(System.lineSeparator())).
                        collect(Collectors.toList());

        if (debug) {
            System.out.println();
            lines.forEach(System.out::println);
            System.out.println();
        }

        return split(lines);
    }

    private List<List<String>> split(List<String> lines) {
        List<List<String>> result = new ArrayList<>();

        int i = 0;
        while (i < lines.size()) {
            String line = lines.get(i);
            if (!line.startsWith("  //")) {
                i++;
                continue;
            }

            int begin = i;
            int end = i;
            while (end < lines.size()) {
                line = lines.get(end);
                if (line.isEmpty() || line.equals("}")) {
                    break;
                }
                end++;
            }

            result.add(lines.subList(begin, end));
            i = end + 1;
        }
        return result;
    }
}

File: basic-proxy-kotlin-impl/src/test/kotlin/com/test/util/PrettyResultBuilder.kt
package com.test.util

import org.objectweb.asm.ClassReader
import org.objectweb.asm.util.TraceClassVisitor
import java.io.ByteArrayOutputStream
import java.io.PrintWriter

class PrettyResultBuilder(val debug: Boolean) {

    fun build(raw: ByteArray): List<List<String>> {
        val outputStream = ByteArrayOutputStream()

        ClassReader(raw).accept(TraceClassVisitor(PrintWriter(outputStream)), 0)
        val lines = outputStream.toString().splitToSequence(System.lineSeparator()).toList()
//        val lines = outputStream.toString().split(System.lineSeparator().toList()

        if (debug) {
            println()
            lines.forEach { println(it) }
            println()
        }

        return split(lines)
    }

    private fun split(lines: List<String>): List<List<String>> {
        val result = mutableListOf<List<String>>()

        var i = 0
        while (i < lines.size) {
            var line = lines[i]
            if (!line.startsWith("  //")) {
                i++
                continue
            }

            val begin = i
            var end = i
            while (end < lines.size) {
                line = lines[end]
                if (line.isEmpty() || line == "}") {
                    break
                }
                end++
            }

            result.add(lines.subList(begin, end))
            i = end + 1
        }
        return result
    }
}